### ISSUE TITLE
System wide Discord push to talk

### ISSUE THREAD

**Issue by baduhai** [getferdi/ferdi#720]
**Is your feature request related to a problem? Please describe.**
Discord only supports push to talk on the web client, while the tab is opened and focused, which means that when Ferdi is minimised, or hidden in the system tray, there is no push to talk within Discord.

**Describe the solution you'd like**
I would like for push to talk to work system-wide, as it does on the Discord desktop client. Discord is currently the only service that I use with its dedicated client, and it'd be nice to ditch it for Ferdi.

What I suggest is that the Discord service override the browser functionality of telling the webpage that it currently has no focus, and that it passes through a selected push to talk key(s), so that Discord believes it is focussed and continues to provide push to talk functionality.

What could also be done is to add an action to the services section of the menubar, it could be called something like *"Push to Talk"* with a checkbox or *"Enable Push to Talk"*. What this action would do is make it so that Ferdi starts listening, even if the Ferdi window has no focus, for the push to talk key(s), pass through said key(s) to Discord, and trick Discord into believing it currently has focus. While this means that the user must configure push to talk twice(once in Discord, and once in Ferdi), it also ensures that Ferdi isn't needlessly listening for a keyboard key while the user doesn't want to use push to talk.

Another way to implement this feature, would be to detect when Discord is listening for a key(s)(I'm actually not sure if this is possible), detected which key(s) Discord is listening for(again, not sure if possible), and then trigger a function so the Ferdi starts listening for said key(s) even if it has no focus(i.e. minimised or in the tray), and trick Discord discord into thinking it's open and has focus.

**Describe alternatives you've considered**
The only current available alternative is using the official Discord.

**Additional context**
I believe this is possible to do, and could be an elegant solution(especially the second alternative) to make Discord push to talk work system wide with Ferdi, though it could very well be against Discord's ToS, in which case that would be a shame.

I was unsure if I should post this feature request here, since I'm not sure how much this project wants to diverge from Franz. If I should have posted it in the Franz repo, since I've never used Franz and I'm unsure of how welcoming they'd be of this feature request. Or if I should have posted it in the recipe repo, since this is after all a feature related to the Discord service. I decided to post here because I'm a Ferdi user, and not a Franz user, and I figured Ferdi has to first support the functionality of tricking the service into thinking it has focus and listening for a push to talk key(s), before the service can actually support it.

If I should post this in either of the other two repos I listed, please tell me. Maybe it's also worth requesting this feature here and in the Franz repo simultaneously, so if the Franz people like the idea, Ferdi is able to get this feature from upstream, and if they don't, hopefully I've made my case convincing enough so you guys can accept it.😊 If I had any idea of how programming with electron and javascript works, I'd make a proof of concept and open a pull request, but I'm a C++ guy and have no idea how these web focused technologies work.

**Comment by obvionaoe** [comment]
I'm not a dev but I don't know if this would be possible as Ferdi uses Discord's web version which doesn't support the push-to-talk feature :neutral_face:

**Comment by baduhai** [comment]
> Discord's web version which doesn't support the push-to-talk feature

In my first comment I explain how one could possibly bypass this.

**Comment by obvionaoe** [comment]
OK, my bad, I though Discord Web didn't even support the feature. I haven't used it in a while xD sorry

**Comment by mahadevans87** [comment]
This is an interesting feature request. Thanks for reporting this. @vantezzen @kytwb @sampathBlam Should we have a mechanism to listen for keyboard hooks and make the services configurable? This could be used by the recipes to listen for the system wide keyboard events and perform functionality inside the recipe itself. What do you think?

**Comment by kris7t** [comment]
Trying to listen for system-wide keyboard events only works under X11. While ferdi runs under Xwayland when started under a Wayland compositor, the compositor won't deliver keyboards events unless its window is in focus. Maybe the push-to-talk listener service could have an (optional) IPC backend (with something like DBus), so Wayland users can bind the push-to-talk key in their compositor (sending an IPC message to ferdi on key push/release).

**Comment by mahadevans87** [comment]
> 
> 
> On Linux, trying to listen for system-wide keyboard events only works under X11. While ferdi runs under Xwayland when started under a Wayland compositor, the compositor won't deliver keyboards events unless its window is in focus. Maybe the push-to-talk listener service could have an (optional) IPC backend (with something like DBus), so Wayland users can bind the push-to-talk key in their compositor (sending an IPC message to ferdi on key push/release).

Yes. An IPC mechanism was in my mind as well. We could leverage electron's main process to pass events to the renderer and thereby to the recipes.
Maybe use something like this library - https://www.npmjs.com/package/iohook ?

**Comment by kris7t** [comment]
At first glance, iohook looks like a library for listening to keyboard evens (on platforms where that's allowed, which is quite possibly everywhere but Wayland. Although I haven't looked into any sandboxing mechanisms recently introduced my Mac OS or Windows, so maybe it's also limited there.) This would be most certainly required for the common cases (unsandboxed Windows or Mac, Linux X11).

I was thinking more along the lines or receiving the push-to-talk even entirely out-of-band, where ferdi does not get access to global keyboard events. For that, DBus came to my mind, because ferdi already has to connect to the session DBus to deliver notifications and display the tray icon. So maybe another, custom DBus interface could be added for "remote control". Admittedly, I am unfamiliar with the process model of Electron, so I don't know how to do that exactly.

**Comment by mahadevans87** [comment]
@kris7t I am starting to work on a small prototype for this. If I make some progress, will put up a draft PR and then probably you can continue up from there? I assume from the other thread are you are busy this week.

**Comment by kris7t** [comment]
@mahadevans87 Great, thanks! 🙏
